File: packages/provide-uterm-go/pty/uidmap.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package pty

import (
	"errors"
	"fmt"
	"os/user"
	"strconv"
	"strings"
)

// ResolvedUser is the OS identity a PTY child is launched as. Port of
// uid_map.ResolvedUser.
type ResolvedUser struct {
	UID   int
	GID   int
	Home  string
	Shell string
	Name  string // OS username (used for supplementary-group resolution)
}

// UidMapError is raised for resolution failures and privilege-policy
// violations. Port of uid_map.UidMapError (a ValueError subclass).
type UidMapError struct{ msg string }

func (e *UidMapError) Error() string { return e.msg }

func newUidMapError(format string, a ...any) *UidMapError {
	return &UidMapError{msg: fmt.Sprintf(format, a...)}
}

// UidMap resolves an application username to an OS (uid, gid, home, shell).
// Port of uid_map.UidMap.
//
// Resolution priority:
//  1. runAsUID (per-session explicit uid)
//  2. runAs (per-session OS username, numeric uid, or "uid:gid")
//  3. table entry keyed on username (same format as runAs; "*" is wildcard)
//  4. os/user lookup on username — user runs as themselves
type UidMap struct {
	table     map[string]string
	allowRoot bool
}

// NewUidMap builds a UidMap. table may be nil.
func NewUidMap(table map[string]string, allowRoot bool) *UidMap {
	if table == nil {
		table = map[string]string{}
	}
	return &UidMap{table: table, allowRoot: allowRoot}
}

// ResolveOpts carries the optional per-session overrides for Resolve. A nil
// *int means "not provided" (mirroring Python's None default).
type ResolveOpts struct {
	RunAs    string // "" means not provided
	RunAsUID *int
	RunAsGID *int
}

// Resolve maps username (+ opts) to a ResolvedUser. Port of UidMap.resolve.
func (m *UidMap) Resolve(username string, opts ResolveOpts) (*ResolvedUser, error) {
	// Validate username early — before touching the name service or the table.
	if username != "" {
		if err := ValidateUsername(username); err != nil {
			return nil, err
		}
	}

	if opts.RunAsUID != nil {
		if err := m.checkPrivilege(*opts.RunAsUID, opts.RunAsGID); err != nil {
			return nil, err
		}
		return m.fromUID(*opts.RunAsUID, opts.RunAsGID)
	}

	if opts.RunAs != "" {
		return m.resolveSpec(opts.RunAs, opts.RunAsGID)
	}

	spec, ok := m.table[username]
	if !ok {
		spec, ok = m.table["*"]
	}
	if ok {
		return m.resolveSpec(spec, opts.RunAsGID)
	}

	u, err := user.Lookup(username)
	if err != nil {
		return nil, newUidMapError("no such OS user: %q", username)
	}
	return m.fromUserRecord(u, opts.RunAsGID)
}

// checkPrivilege enforces the no-root policy unless allowRoot is set. Port of
// UidMap._check_privilege.
func (m *UidMap) checkPrivilege(uid int, gid *int) error {
	if m.allowRoot {
		return nil
	}
	if uid == 0 || (gid != nil && *gid == 0) {
		gidRepr := "None"
		if gid != nil {
			gidRepr = strconv.Itoa(*gid)
		}
		return newUidMapError("resolving to privileged %d:%s is not allowed", uid, gidRepr)
	}
	return nil
}

// fromUID resolves an explicit uid (+ optional gid). Port of UidMap._from_uid.
// An unknown uid yields a synthetic user rooted at "/" with /bin/sh.
func (m *UidMap) fromUID(uid int, gid *int) (*ResolvedUser, error) {
	if err := m.checkPrivilege(uid, gid); err != nil {
		return nil, err
	}
	u, err := user.LookupId(strconv.Itoa(uid))
	if err != nil {
		resolvedGID := uid
		if gid != nil {
			resolvedGID = *gid
		}
		return &ResolvedUser{
			UID:   uid,
			GID:   resolvedGID,
			Home:  "/",
			Shell: defaultShell,
			Name:  strconv.Itoa(uid),
		}, nil
	}
	pwUID, _ := strconv.Atoi(u.Uid)
	pwGID, _ := strconv.Atoi(u.Gid)
	resolvedGID := pwGID
	if gid != nil {
		resolvedGID = *gid
	}
	return &ResolvedUser{
		UID:   pwUID,
		GID:   resolvedGID,
		Home:  u.HomeDir,
		Shell: lookupShell(pwUID, u.Username),
		Name:  u.Username,
	}, nil
}

// resolveSpec parses "OS-username" | "uid" | "uid:gid". Port of
// UidMap._resolve_spec.
func (m *UidMap) resolveSpec(spec string, runAsGID *int) (*ResolvedUser, error) {
	if strings.Contains(spec, ":") {
		parts := strings.SplitN(spec, ":", 2)
		uid, err := strconv.Atoi(parts[0])
		if err != nil {
			return nil, fmt.Errorf("invalid uid in spec %q: %w", spec, err)
		}
		gid, err := strconv.Atoi(parts[1])
		if err != nil {
			return nil, fmt.Errorf("invalid gid in spec %q: %w", spec, err)
		}
		return m.fromUID(uid, &gid)
	}

	if uid, err := strconv.Atoi(spec); err == nil {
		return m.fromUID(uid, runAsGID)
	}

	u, err := user.Lookup(spec)
	if err != nil {
		return nil, newUidMapError("no such OS user: %q", spec)
	}
	return m.fromUserRecord(u, runAsGID)
}

// fromUserRecord builds a ResolvedUser from an os/user record, applying the
// gid override + root-privilege check (shared by name-based resolution paths).
func (m *UidMap) fromUserRecord(u *user.User, runAsGID *int) (*ResolvedUser, error) {
	pwUID, _ := strconv.Atoi(u.Uid)
	pwGID, _ := strconv.Atoi(u.Gid)
	gid := pwGID
	if runAsGID != nil {
		gid = *runAsGID
	}
	if err := m.checkPrivilege(pwUID, &gid); err != nil {
		return nil, err
	}
	return &ResolvedUser{
		UID:   pwUID,
		GID:   gid,
		Home:  u.HomeDir,
		Shell: lookupShell(pwUID, u.Username),
		Name:  u.Username,
	}, nil
}

// IsUidMapError reports whether err is (or wraps) a *UidMapError.
func IsUidMapError(err error) bool {
	var target *UidMapError
	return errors.As(err, &target)
}

File: packages/provide-uterm-ts/src/pty/uid-map.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * Which operating-system user a session runs as.
 *
 * Port of `provide.uterm.pty.uid_map`. This decides the identity a shell is
 * started under, so the rule that matters is the one it refuses: nothing
 * resolves to uid 0 or gid 0 unless an operator has explicitly allowed root.
 * That is the difference between a session running as a person and one running
 * as the machine, and it is checked on every path in — an explicit uid, a
 * run-as spec, a table entry, and the user's own passwd record.
 *
 * The passwd database is the caller's: a lookup is injected rather than
 * imported, so this module runs anywhere and a test can describe a machine
 * rather than depend on one.
 */

import { pyInt, pyRepr } from "../pycompat/index.ts";
import { validateUsername } from "./validate.ts";

/** What a session needs to know about the user it runs as. */
export interface ResolvedUser {
  uid: number;
  gid: number;
  home: string;
  shell: string;
  name: string;
}

/** One row of the passwd database. */
export interface PasswdEntry {
  name: string;
  uid: number;
  gid: number;
  home: string;
  shell: string;
}

/** How this module reads the passwd database. */
export interface PasswdLookup {
  /** By name, or nothing when there is no such user. */
  byName(name: string): PasswdEntry | undefined;
  /** By number, or nothing when no user has it. */
  byUid(uid: number): PasswdEntry | undefined;
}

/** A resolution that cannot be allowed, or cannot be made at all. */
export class UidMapError extends Error {}

/** What a caller may say about how a session resolves. */
export interface UidMapOptions {
  /** Application username → run-as spec. `*` matches anybody. */
  table?: Readonly<Record<string, string>>;
  /**
   * Whether root may be resolved to.
   *
   * Off by default, and the default is the point: a deployment that means to
   * run sessions as root has to say so.
   */
  allowRoot?: boolean;
  passwd: PasswdLookup;
}

/** What one resolution may override. */
export interface ResolveOptions {
  /** An OS username, a numeric uid, or `uid:gid`. */
  runAs?: string;
  /** An explicit uid, which beats everything else. */
  runAsUid?: number;
  /** An explicit gid, used where the chosen path does not name one. */
  runAsGid?: number;
}

/** How a uid nobody has is described, having nothing to describe it with. */
const UNKNOWN_UID_HOME = "/";
const UNKNOWN_UID_SHELL = "/bin/sh";

/**
 * Resolve an application username to an operating-system identity.
 *
 * In order: an explicit uid, a run-as spec, a table entry for this user or the
 * wildcard, and finally the user's own passwd record — so a session runs as
 * itself unless something says otherwise.
 */
export class UidMap {
  readonly #table: Readonly<Record<string, string>>;
  readonly #allowRoot: boolean;
  readonly #passwd: PasswdLookup;

  constructor(options: UidMapOptions) {
    this.#table = options.table ?? {};
    this.#allowRoot = options.allowRoot ?? false;
    this.#passwd = options.passwd;
  }

  /**
   * The identity this session runs as.
   *
   * @throws {UidMapError} When there is no such user, or when the answer would
   *   be privileged and root has not been allowed.
   * @throws {Error} When the username is not one an operating system would
   *   accept.
   */
  resolve(username: string, options: ResolveOptions = {}): ResolvedUser {
    // Checked before anything is looked up: a name with a null byte or a colon
    // in it must not reach a passwd lookup at all.
    if (username !== "") {
      validateUsername(username);
    }

    // An explicit uid beats everything else, including a run-as naming a
    // different user. The privilege check lives in `#fromUid`, which every
    // path in goes through.
    if (options.runAsUid !== undefined) {
      return this.#fromUid(options.runAsUid, options.runAsGid);
    }

    if (options.runAs !== undefined) {
      return this.#fromSpec(options.runAs, options.runAsGid);
    }

    const spec = this.#table[username] || this.#table["*"];
    if (spec !== undefined) {
      return this.#fromSpec(spec, options.runAsGid);
    }

    const entry = this.#passwd.byName(username);
    if (entry === undefined) {
      throw new UidMapError(`no such OS user: ${pyRepr(username)}`);
    }
    const gid = options.runAsGid ?? entry.gid;
    this.#checkPrivilege(entry.uid, gid);
    return { uid: entry.uid, gid, home: entry.home, shell: entry.shell, name: entry.name };
  }

  /**
   * Refuse a privileged answer.
   *
   * Either half counts: a session running as an unprivileged user in group
   * zero is still in group zero.
   */
  #checkPrivilege(uid: number, gid: number | undefined): void {
    if (this.#allowRoot) {
      return;
    }
    if (uid === 0 || gid === 0) {
      throw new UidMapError(`resolving to privileged ${uid}:${gid === undefined ? "None" : gid} is not allowed`);
    }
  }

  /**
   * The identity behind a number.
   *
   * A uid nobody has still resolves — to itself, at `/`, with a plain shell.
   * The privilege check has already run, so this cannot be root.
   */
  #fromUid(uid: number, gid: number | undefined): ResolvedUser {
    this.#checkPrivilege(uid, gid);
    const entry = this.#passwd.byUid(uid);
    if (entry === undefined) {
      return {
        uid,
        gid: gid ?? uid,
        home: UNKNOWN_UID_HOME,
        shell: UNKNOWN_UID_SHELL,
        name: String(uid),
      };
    }
    return { uid: entry.uid, gid: gid ?? entry.gid, home: entry.home, shell: entry.shell, name: entry.name };
  }

  /** A run-as spec: an OS username, a `uid`, or a `uid:gid`. */
  #fromSpec(spec: string, runAsGid: number | undefined): ResolvedUser {
    if (spec.includes(":")) {
      const separator = spec.indexOf(":");
      const uid = pyInt(spec.slice(0, separator));
      const gid = pyInt(spec.slice(separator + 1));
      if (uid === undefined || gid === undefined) {
        throw new UidMapError(`invalid literal for int() with base 10: ${pyRepr(spec.slice(separator + 1))}`);
      }
      return this.#fromUid(uid, gid);
    }

    const uid = pyInt(spec);
    if (uid !== undefined) {
      return this.#fromUid(uid, runAsGid);
    }

    const entry = this.#passwd.byName(spec);
    if (entry === undefined) {
      throw new UidMapError(`no such OS user: ${pyRepr(spec)}`);
    }
    const gid = runAsGid ?? entry.gid;
    this.#checkPrivilege(entry.uid, gid);
    return { uid: entry.uid, gid, home: entry.home, shell: entry.shell, name: entry.name };
  }
}
